library/pkcs7_write.c

```diff
@@ -0,0 +1,843 @@
+        }
+        mbedtls_free( hash );
+        if( ret )
+            mbedtls_printf( "Failed to add signature to PKCS7\n" );
+
+        return ( ret );
+    }
+
+    /* if here we know that `priv` contains a private key, not a signature */
+    priv_key = mbedtls_calloc( 1, sizeof( *priv_key ) );
+    if( priv_key == NULL )
+        return ( MBEDTLS_ERR_PKCS7_ALLOC_FAILED );
+
+    mbedtls_pk_init( priv_key );
+    /* make sure private key parses into private key format */
+    ret = mbedtls_pk_parse_key( priv_key, priv, priv_size, NULL, 0,
+                                pkcs7_info->rng_func, pkcs7_info->rng_param );
+    if( ret != 0 )
+        goto out;
+
+    /* make sure private key is matched with public key */
+    ret = mbedtls_pk_check_pair( &( pub->pk ), priv_key,
+                                 pkcs7_info->rng_func, pkcs7_info->rng_param );
+    if( ret != 0 )
+        goto out;
+
+    /* make sure private key is RSA, otherwise quit */
+    sig_type = (char *) priv_key->pk_info->name;
+    if( strcmp( sig_type, "RSA" ) ) {
+        ret = MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO;
+        goto out;
+    }
+    /* get size of RSA signature, ex 2048, 4096 ... */
+    sig_size_bits = priv_key->pk_info->get_bitlen( priv_key->pk_ctx );
+
+    signature = mbedtls_calloc( 1, sig_size_bits / 8 );
+    if( signature == NULL ) {
+        ret = MBEDTLS_ERR_PKCS7_ALLOC_FAILED;
+        goto out;
+    }
+
+
+    ret = mbedtls_pk_sign( priv_key, pkcs7_info->hash_funct,
+                           (const unsigned char *) hash, hash_size,
+                           (unsigned char *) signature, sig_size_bits / 8,
+                           &sig_size, pkcs7_info->rng_func, pkcs7_info->rng_param );
+    if( ret != 0 ) {
+        mbedtls_free( signature );
+        goto out;
+    }
+    ret = pkcs7_write_data( start, size, ptr, WRITE_SIGNATURE, signature,
+                            sig_size, 0 );
+    mbedtls_free( signature );
+
+    if( ret != 0 ) {
+        mbedtls_printf(
+            "Failed to add signature to PKCS7 (signature generation was successful "
+            "however)\n" );
+    }
+out:
+    mbedtls_pk_free( priv_key );
+    mbedtls_free( priv_key );
+    mbedtls_free( hash );
+
+    return ( ret );
+}
+
+static int pkcs7_set_algorithm_ids( unsigned char **start, size_t *size,
+                                    unsigned char **ptr, mbedtls_pkcs7_info *pkcs7_info,
+                                    mbedtls_x509_crt *pub, unsigned char *priv,
+                                    size_t priv_size )
+{
+    int ret;
+    char *sig_type = NULL;
+
+    
+    ret = pkcs7_set_signature( start, size, ptr, pkcs7_info, pub, priv, priv_size );
+    if( ret != 0 )
+        return ( ret );
+
+    /* make sure it is rsa encryption, that is all we support right now */
+    sig_type = (char *) pub->pk.pk_info->name;
+    if( strcmp( sig_type, "RSA" ) ) {
+        ret = MBEDTLS_ERR_PKCS7_INVALID_CERT;
```

Comment:
Having distinct errors for unsupported vs invalid is nice, but it's also problematic because it creates an expectation that the implementation always distinguishes them, even whe. I made them different error codes in the PSA crypto API, but in retrospect I think I made a mistake: it's led to complaints about things like getting a not-supported error when asking for a PSS signature with an ECC key in a build without RSA (so it sees PSS as the signature type and says “not supported”, having no idea what key type PSS would be compatible with).

Note that I'm not saying it's a bad idea to distinguish them _here_, just that it isn't always a good idea. Distinguishing not-supported from invalid does have advantages when debugging. Typically, it's advantage if you known your code better than your data (it gives you more feedback about what's going wrong in a deployment), but it's a problem if you known your data better than your code (the biggest complaints due to the separate error codes are from compliance testing).